### ISSUE TITLE
Remove Task.Run from DefaultHealthCheckService checking services

### ISSUE THREAD

**Issue by PetrovskiAlex** [dotnet/aspnetcore#41758]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Hi guys, I was reading the health check source code and came across this line of code

tasks[index++] = Task.Run(() => RunCheckAsync(registration, cancellationToken), cancellationToken);

https://github.com/dotnet/aspnetcore/blob/bf50999cd8e60d4c683b772dffa9898794a8433a/src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs#L56

But, I am a bit confused, why do need to wrap the RunCkeckAsync task to another Task.Run? Could someone explain to me please what bad things happen if we just remove Task.Run? What is the difference from this code 

tasks[i] = RunPublisherAsync(publishers[i], report, cancellationToken);

https://github.com/dotnet/aspnetcore/blob/bf50999cd8e60d4c683b772dffa9898794a8433a/src/HealthChecks/HealthChecks/src/HealthCheckPublisherHostedService.cs#L160

### Describe the solution you'd like

It seems, we safely could just remove Task.Run and invoke RunCheckAsync directly, so we decrease thread pool pressure and everything will work fine.

**Comment by davidfowl** [comment]
It's done to force the async offload to happen before the first await. Otherwise, if the health check runs synchronously there will be less parallelism.

**Comment by PetrovskiAlex** [comment]
@davidfowl thank you for your time. I have more questions after your response :) 
Why does the framework(aspnet) suppose that health check runs synchronously? I assume most of the health checks run asynchronously as they are suppose to(as we can see on the signature of the CheckAsync method). In our projects we have 3-5 health checks, and no one is synchronous
If an app has a sync health check, an app can wrap only that health check to Task.Run.

**Comment by ovska** [comment]
The documentation for `IHealthCheck` mentions databases, which should be async, and application internal state, which I'd imagine are very low overhead even if synchronous. Does the reasoning have to do with the fact that the check durations are being measured?

I have to agree with Alex, seems weird to assume the worst when the interface returns a Task.

**Comment by davidfowl** [comment]
I don't know what there is to disagree with, is there a real problem here or is this about reading the code and trying to understand it?

**Comment by PetrovskiAlex** [comment]
@davidfowl Yes, you are right, it is about reading the code and trying to understand why the author wrote Task.Run. I did not measure the performance of our apps and I guess there is no a problem with health checks. If we have nothing to disagree with, why would not do that? :)

**Comment by davidfowl** [comment]
> It seems, we safely could just remove Task.Run and invoke RunCheckAsync directly, so we decrease thread pool pressure and everything will work fine.

There's an assertion that there is a problem. The code is preemptively offloading to the threadpool, which is a fine pattern. Usually running the code directly vs dispatching has to do with:
1. The likely hood of it returning synchronously or blocking.
1. The amount of code that runs synchronously before the await happens.

In this case, I think point 2 is why there's dispatch. I don't see a problem with the current logic besides it being a potentially little less efficient if the tasks are already completed.

**Comment by PetrovskiAlex** [comment]
@davidfowl thanks a lot for your responses. 
What does the phrase exactly mean - "the amount of code"? Why it should not depend on code?

1) logger.Log("Info");
2) Thread.Sleep(1000);
These two lines of code have a significantly different impact on the system

And if we look at several health checks:
- DbConnection - https://github.com/dotnet/aspnetcore/blob/bf50999cd8e60d4c683b772dffa9898794a8433a/src/Middleware/HealthChecks/test/testassets/HealthChecksSample/DbConnectionHealthCheck.cs#L39
- DbContext https://github.com/dotnet/aspnetcore/blob/bf50999cd8e60d4c683b772dffa9898794a8433a/src/Middleware/HealthChecks.EntityFrameworkCore/src/DbContextHealthCheck.cs#L35
- Redis https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/c60573659a5e415fbbf0fda8363ca798b32890f4/src/HealthChecks.Redis/RedisHealthCheck.cs#L23
- MySql https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/c60573659a5e415fbbf0fda8363ca798b32890f4/src/HealthChecks.MySql/MySqlHealthCheck.cs#L21
- ElasticSearch https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/c60573659a5e415fbbf0fda8363ca798b32890f4/src/HealthChecks.Elasticsearch/ElasticsearchHealthCheck.cs#L18
- Kafka https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/c60573659a5e415fbbf0fda8363ca798b32890f4/src/HealthChecks.Kafka/KafkaHealthCheck.cs#L18

we can see that most of them have a small "amount of code".
In addition, most of the books say - do not waste threads, this fundamental truth is gonna break for me :)

**Comment by davidfowl** [comment]
It's really about this code https://github.com/dotnet/aspnetcore/blob/bf50999cd8e60d4c683b772dffa9898794a8433a/src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs#L78-L105. Also no threads are being wasted...

**Comment by PetrovskiAlex** [comment]
@davidfowl Could you point me please to that piece of code, what takes too long time, and that the reason why we should dispatch?
As I know Task.Run queue task to a TaskScheduler.Default, and in aspnet is ThreadPoolScheduler. Is not wasting threads of ThreadPool?

**Comment by davidfowl** [comment]
> @davidfowl Could you point me please to that piece of code, what takes too long time, and that the reason why we should dispatch?

to force health checks to be executed in parallel.

> As I know Task.Run queue task to a TaskScheduler.Default, and in aspnet is ThreadPoolScheduler. Is not wasting threads of ThreadPool?

Running a callback on the thread pool is queuing a work item and then subsequently using and existing thread to execute the continuation.

**Comment by PetrovskiAlex** [comment]
@davidfowl thank you for your time, it helped me a lot to understand why it was written that way
